README.md
DoubleBagFTPS
=============

DoubleBagFTPS extends the core Net::FTP class to provide implicit and explicit FTPS support.

Install
-------

    $ [sudo] gem install double-bag-ftps

**Note**: Your Ruby installation must have OpenSSL support.

Usage
-----
    require 'double_bag_ftps'

Example 1:

    # Connect to a host using explicit FTPS and do not verify the host's cert
    ftps = DoubleBagFTPS.new
    ftps.ssl_context = DoubleBagFTPS.create_ssl_context(:verify_mode => OpenSSL::SSL::VERIFY_NONE)
    ftps.connect('some host')
    ftps.login('usr', 'passwd')

Example 2:

    DoubleBagFTPS.open('host', 'usr', 'passwd', nil, DoubleBagFTPS::IMPLICIT) do |ftps|
      ...
    end

Interface
---------

    # Constants used for setting FTPS mode
    DoubleBagFTPS::EXPLICIT
    DoubleBagFTPS::IMPLICIT

    DoubleBagFTPS.new(host = nil, user = nil, passwd = nil, acct = nil, ftps_mode = EXPLICIT, ssl_context_params = {})
    DoubleBagFTPS.open(host, user = nil, passwd = nil, acct = nil, ftps_mode = EXPLICIT, ssl_context_params = {})

    # Returns an OpenSSL::SSL::SSLContext using params to set set the corresponding SSLContext attributes.
    DoubleBagFTPS.create_ssl_context(params = {})

    # Set the FTPS mode to implicit (DoubleBagFTPS::IMPLICIT) or explicit (DoubleBagFTPS::EXPLICIT).
    # The default FTPS mode is explicit. 
    ftps_mode=(ftps_mode)

    # Same as Net::FTP.connect, but will use port 990 when using implicit FTPS and a port is not specified.
    connect(host, port = ftps_implicit? ? IMPLICIT_PORT : FTP_PORT)

    # Same as Net::FTP.login, but with optional auth param to control the value that is sent with the AUTH command.
    login(user = 'anonymous', passwd = nil, acct = nil, auth = 'TLS')

    ftps_explicit?
    ftps_implicit?

More Information
----------------

* [Net::FTP RDoc](http://ruby-doc.org/stdlib/libdoc/net/ftp/rdoc/index.html)
* [OpenSSL RDoc](http://ruby-doc.org/stdlib/libdoc/openssl/rdoc/index.html)

License
-------
Copyright © 2011, Bryan Nix. DoubleBagFTPS is released under the MIT license. See LICENSE file for details.

CHANGELOG.md
## 0.1.4

Buffixes:

- Fix active mode transfers hanging in some versions of Ruby 2.1

## 0.1.3

Bugfixes:

- Fix active mode transfers hanging in some versions of Ruby 2.2 and
  2.3

## 0.1.2

Bugfixes:

  - Reuse SSL session when creating new sockets

## 0.1.1

Features:

  - Ruby 2.0-p0 compatibility

Bugfixes:

  - Active mode data transfers properly establish a connection


spec/spec_helper.rb
$:.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))
require 'double_bag_ftps'

require 'tmpdir'

HOST   = 'ftp.secureftp-test.com'
USR    = 'test'
PASSWD = 'test'


spec/double_bag_ftps_spec.rb
require File.expand_path(File.dirname(__FILE__) + '/spec_helper')

shared_examples_for "DoubleBagFTPS" do
  it "connects to a remote host" do
    lambda {@ftp.connect(HOST)}.should_not raise_error
  end

  it "logs in with a user name and password" do
    @ftp.connect(HOST)
    lambda {@ftp.login(USR, PASSWD)}.should_not raise_error
  end

  it "can open a secure data channel" do
    @ftp.connect(HOST)
    @ftp.login(USR, PASSWD)
    @ftp.send(:transfercmd, 'nlst').should be_an_instance_of OpenSSL::SSL::SSLSocket
  end

  it "can retrieve a file" do
    Dir.mktmpdir do |temp_dir|
      @ftp.connect HOST
      @ftp.login USR, PASSWD
      filename = @ftp.nlst.first
      filename.should_not be_nil
      local_path = File.join(temp_dir, filename)
      @ftp.get(filename, local_path)
      File.exists?(local_path).should be_true
    end
  end

  it "prevents setting the FTPS mode while connected" do
    @ftp.connect(HOST)
    lambda {@ftp.ftps_mode = DoubleBagFTPS::IMPLICIT}.should raise_error RuntimeError
  end

  it "prevents setting the FTPS mode to an unrecognized value" do
    lambda {@ftp.ftps_mode = 'dummy value'}.should raise_error ArgumentError
  end

end

describe DoubleBagFTPS do
  context "implicit" do
    before(:each) do
      @ftp = DoubleBagFTPS.new
      @ftp.ftps_mode = DoubleBagFTPS::IMPLICIT
      @ftp.passive = true
      @ftp.ssl_context = DoubleBagFTPS.create_ssl_context(:verify_mode => OpenSSL::SSL::VERIFY_NONE)
    end

    after(:each) do
      @ftp.close unless @ftp.welcome.nil?
    end

    it "uses an SSLSocket when first connected" do
      @ftp.connect(HOST)
      @ftp.instance_eval {def socket; @sock; end}
      @ftp.socket.should be_an_instance_of OpenSSL::SSL::SSLSocket
    end

    it_should_behave_like "DoubleBagFTPS"
  end

  context "explicit" do
    before(:each) do
      @ftp = DoubleBagFTPS.new
      @ftp.ftps_mode = DoubleBagFTPS::EXPLICIT
      @ftp.passive = true
      @ftp.ssl_context = DoubleBagFTPS.create_ssl_context(:verify_mode => OpenSSL::SSL::VERIFY_NONE)
    end

    after(:each) do
      @ftp.close unless @ftp.welcome.nil?
    end

    it "does not use an SSLSocket when first connected" do
      @ftp.connect(HOST)
      @ftp.instance_eval {def socket; @sock; end}
      @ftp.socket.should_not be_an_instance_of OpenSSL::SSL::SSLSocket
    end

    it_should_behave_like "DoubleBagFTPS"
  end
end


lib/double_bag_ftps.rb
require 'net/ftp'
require 'rubygems'
begin
  require 'openssl'
rescue LoadError
end

class DoubleBagFTPS < Net::FTP
  EXPLICIT = :explicit
  IMPLICIT = :implicit
  IMPLICIT_PORT = 990

  # The form of FTPS that should be used. Either EXPLICIT or IMPLICIT.
  # Defaults to EXPLICIT.
  attr_reader :ftps_mode

  # The OpenSSL::SSL::SSLContext to use for creating all OpenSSL::SSL::SSLSocket objects.
  attr_accessor :ssl_context

  def initialize(host = nil, user = nil, passwd = nil, acct = nil, ftps_mode = EXPLICIT, ssl_context_params = {})
    raise ArgumentError unless valid_ftps_mode?(ftps_mode)
    @ftps_mode = ftps_mode
    @ssl_context = DoubleBagFTPS.create_ssl_context(ssl_context_params)
    super(host, user, passwd, acct)
  end

  def DoubleBagFTPS.open(host, user = nil, passwd = nil, acct = nil, ftps_mode = EXPLICIT, ssl_context_params = {})
    if block_given?
      ftps = new(host, user, passwd, acct, ftps_mode, ssl_context_params)
      begin
        yield ftps
      ensure
        ftps.close
      end
    else
      new(host, user, passwd, acct, ftps_mode, ssl_context_params)
    end
  end

  #
  # Allow @ftps_mode to be set when @sock is not connected
  #
  def ftps_mode=(ftps_mode)
    # Ruby 1.8.7/1.9.2 compatible check
    if (defined?(NullSocket) && @sock.kind_of?(NullSocket)) || @sock.nil? || @sock.closed?
      raise ArgumentError unless valid_ftps_mode?(ftps_mode)
      @ftps_mode = ftps_mode
    else
      raise 'Cannot set ftps_mode while connected'
    end
  end

  #
  # Establishes the command channel.
  # Override parent to record host name for verification, and allow default implicit port.
  #
  def connect(host, port = ftps_implicit? ? IMPLICIT_PORT : FTP_PORT)
    @hostname = host
    super
  end

  def login(user = 'anonymous', passwd = nil, acct = nil, auth = 'TLS')
    if ftps_explicit?
      synchronize do
        sendcmd('AUTH ' + auth) # Set the security mechanism
        @sock = ssl_socket(@sock)
      end
    end
    
    super(user, passwd, acct)
    voidcmd('PBSZ 0') # The expected value for Protection Buffer Size (PBSZ) is 0 for TLS/SSL 
    voidcmd('PROT P') # Set data channel protection level to Private
  end

  #
  # Override parent to allow an OpenSSL::SSL::SSLSocket to be returned
  # when using implicit FTPS
  #
  def open_socket(host, port, defer_implicit_ssl = false)
    if defined? SOCKSSocket and ENV["SOCKS_SERVER"]
      @passive = true
      sock = SOCKSSocket.open(host, port)
    else
      sock = TCPSocket.open(host, port)
    end
    return (!defer_implicit_ssl && ftps_implicit?) ? ssl_socket(sock) : sock
  end
  private :open_socket

  #
  # Override parent to support ssl sockets
  #
  def transfercmd(cmd, rest_offset = nil)
    if @passive
      host, port = makepasv

      if @resume and rest_offset
        resp = sendcmd('REST ' + rest_offset.to_s)
        if resp[0] != ?3
          raise FTPReplyError, resp
        end
      end
      conn = open_socket(host, port, true)
      resp = sendcmd(cmd)
      # skip 2XX for some ftp servers
      resp = getresp if resp[0] == ?2
      if resp[0] != ?1
        raise FTPReplyError, resp
      end
      conn = ssl_socket(conn) # SSL connection now possible after cmd sent
    else
      sock = makeport
      sendport(sock.addr[3], sock.addr[1]) if sendport_needed?
      if @resume and rest_offset
        resp = sendcmd('REST ' + rest_offset.to_s)
        if resp[0] != ?3
          raise FTPReplyError, resp
        end
      end
      resp = sendcmd(cmd)
      # skip 2XX for some ftp servers
      resp = getresp if resp[0] == ?2
      if resp[0] != ?1
        raise FTPReplyError, resp
      end
      conn = sock.accept
      conn = ssl_socket(conn)
      sock.close
    end
    return conn
  end
  private :transfercmd

  # Before ruby-2.2.3, makeport called sendport automatically.  After
  # Ruby 2.3.0, makeport does not call sendport automatically, so do
  # it ourselves.  This change to Ruby's FTP lib has been backported
  # to Ruby 2.1 since version 2.1.7.
  def sendport_needed?
    @sendport_needed ||= begin
      Gem::Version.new(RUBY_VERSION) >= Gem::Version.new("2.2.3") ||
      Gem::Version.new(RUBY_VERSION) >= Gem::Version.new("2.1.7") &&
      Gem::Version.new(RUBY_VERSION) < Gem::Version.new("2.2.0")
    end
  end
  private :sendport_needed?

  def ftps_explicit?; @ftps_mode == EXPLICIT end
  def ftps_implicit?; @ftps_mode == IMPLICIT end

  def valid_ftps_mode?(mode)
    mode == EXPLICIT || mode == IMPLICIT
  end
  private :valid_ftps_mode?

  #
  # Returns a connected OpenSSL::SSL::SSLSocket
  #
  def ssl_socket(sock)
    raise 'SSL extension not installed' unless defined?(OpenSSL)
    sock = OpenSSL::SSL::SSLSocket.new(sock, @ssl_context)
    if @ssl_session
      sock.session = @ssl_session
    end
    sock.sync_close = true
    sock.connect
    print "get: #{sock.peer_cert.to_text}" if @debug_mode
    unless @ssl_context.verify_mode == OpenSSL::SSL::VERIFY_NONE
      sock.post_connection_check(@hostname)
    end
    @ssl_session = sock.session
    decorate_socket sock
    return sock
  end
  private :ssl_socket

  # Ruby 2.0's Ftp class closes sockets by first doing a shutdown,
  # setting the read timeout, and doing a read.  OpenSSL doesn't
  # have those methods, so fake it.
  #
  # Ftp calls #close in an ensure block, so the socket will still get
  # closed.

  def decorate_socket(sock)

    def sock.shutdown(how)
      @shutdown = true
    end

    def sock.read_timeout=(seconds)
    end

    # Skip read after shutdown.  Prevents 2.0 from hanging in
    # Ftp#close

    def sock.read(*args)
      return if @shutdown
      super(*args)
    end

  end
  private :decorate_socket

  def DoubleBagFTPS.create_ssl_context(params = {})
    raise 'SSL extension not installed' unless defined?(OpenSSL)
    context = OpenSSL::SSL::SSLContext.new
    context.set_params(params)
    return context
  end

end
